### PR TITLE
common: Make ThreadId (absl::hash)-able

### PR DESCRIPTION
**Pull request by auni53**
Signed-off-by: Auni Ahsan <auni@google.com>

Description: I want to use ThreadIds as keys for a map for my use case.
Risk Level: Low I hope, but I have no idea how to build or test the ThreadIdImplWin32 class.
Testing: Added a test/common/common/thread_impl_test.cc that does a very basic hash/lookup test.

### CODE REVIEW

**Comment by auni53** [comment]
came from discussion w @curiouserrandy @mergeconflict @htuch

File: include/envoy/thread/thread.h [hunk 1]
```diff
@@ -14,6 +14,7 @@ class ThreadId {
 public:
   virtual ~ThreadId() {}
 
+  virtual int tid() const PURE;
```
Comment (mergeconflict) [inline include/envoy/thread/thread.h]: This needs to be int64_t so it's not truncated on posix.

**Comment by htuch** [comment]
@sesmith177 does this work on Windows? Is the DWORD a unique identifier of the thread on Windows?

**Comment by jmarantz** [comment]
Please read this:

https://linux.die.net/man/3/pthread_self

And in particular:

> POSIX.1 allows an implementation wide freedom in choosing the type used to represent a thread ID; for example, representation using either an arithmetic type or a structure is permitted. Therefore, variables of type pthread_t can't portably be compared using the C equality operator (==); use pthread_equal(3) instead.
> Thread identifiers should be considered opaque: any attempt to use a thread ID other than in pthreads calls is nonportable and can lead to unspecified results.
> 
> Thread IDs are only guaranteed to be unique within a process. A thread ID may be reused after a terminated thread has been joined, or a detached thread has terminated.
> 
> The thread ID returned by pthread_self() is not the same thing as the kernel thread ID returned by a call to gettid(2).

**Comment by jmarantz** [comment]
Also, I think copying thread IDs is an operation you could add, but you may be able to just share references or use shared_ptr.

**Comment by auni53** [comment]
Agreed on test coverage, tbh I'm not sure on what the invariants should be. As well, there's no ThreadIdImplTest; should I add one?

@jmarantz will respond to your first post later, need to read it in more detail to understand it. I did recall it wasn't an issue of copy/moving, but rather if you pull a ThreadId out of the ThreadFactory that gets created as a temporary object which was causing problems. I suppose I could wrap that in a smart pointer and store it, but seems odd to me when I only need a unique integer representation. (My implementation just uses the debugString() right now.)

**Comment by jmarantz** [comment]
Basically it says "you can't assume anything about the rep for a thread ID". Only valid operation is the functional compare for equality.

**Comment by curiouserrandy** [comment]
@jmarantz : The requirement this PR is responding to is "We need a unique per-thread integer identifier for lookup purposes".  Given that, I'm inclined to just change the name of the method in response to your concerns.  Would uniqueId() satisfy those concerns?  Then figuring out how to create a unique id could be done on a per-implementation basis.  (In all implementations I'm aware of so far it would just be a cast.)

@mergeconflict : My only requirements were "per-thread unique", and I'm not sure how to test that other than creating a lot of threads and making sure there aren't any collisions, which as far as tests go is a bit ... lacking :-}.  Any other ideas?

**Comment by jmarantz** [comment]
How about using debugString() as the map key, or if that doesn't feel write, add new 'std::string mapKey()' method whose impl in posix may just return debugString()?

**Comment by curiouserrandy** [comment]
This effort was started because I strongly objected to relying on debugString() for the uniqueness property (or, indeed, for any property that needed to be true in production code).  I would have no objection to the mapKey() method to solve this problem.

**Comment by mergeconflict** [comment]
> @jmarantz : The requirement this PR is responding to is "We need a unique per-thread integer identifier for lookup purposes". Given that, I'm inclined to just change the name of the method in response to your concerns. Would uniqueId() satisfy those concerns? Then figuring out how to create a unique id could be done on a per-implementation basis. (In all implementations I'm aware of so far it would just be a cast.)
> 
> @mergeconflict : My only requirements were "per-thread unique", and I'm not sure how to test that other than creating a lot of threads and making sure there aren't any collisions, which as far as tests go is a bit ... lacking :-}. Any other ideas?

How about declaring `ThreadId::operator==` and `std::hash<ThreadId>`?

**Comment by jmarantz** [comment]
@mergeconflict I like it. That sounds more efficient than returning a string key.

**Comment by htuch** [comment]
Yep, that sounds good to me as well.

**Comment by curiouserrandy** [comment]
+1

**Comment by auni53** [comment]
Ummmm defining specializations for std::hash seems to be prohibited by Google C++, [if I'm reading this correctly](https://google.github.io/styleguide/cppguide.html#std_hash). Instead I'll define absl::Hash which is apparently safer.

**Comment by jmarantz** [comment]
What's the status of this?

For the record I vote for not using a specialization for std::hash *or* absl::Hash, but instead defining a hash/compare functor, passed as template parameters 3 and 4 to absl::flat_hash_map.  See me if this is not clear. I think this should be very straightforward to do.

**Comment by auni53** [comment]
My bad, this is still on my radar but I got pulled into other things! Will post another PR with the right absl::hash implementation soon.

**Comment by jmarantz** [comment]
You don't need another PR -- I just re-opened this one so there is comment continuity. I think this is pretty simple; you just need a `using` declaration pointing to a hash() method you add to ThreadId.

**Pull request closed without merging**